packages/amplify-provider-awscloudformation/src/pre-push-cfn-processor/cfn-pre-processor.ts

```diff
@@ -24,3 +39,176 @@ export async function preProcessCFNTemplate(filePath: string): Promise<string> {
+    customExecutionPolicy = customExecutionPolicyForFunction;
+  }
+  if (service === 'ElasticContainer') {
+    const roleName = cfnTemplate.Resources.TaskDefinition.Properties.ExecutionRoleArn["Fn::GetAtt"][0];
+    customExecutionPolicy = customExecutionPolicyForContainer;
+    const role = Fn.Ref(roleName);
+    customExecutionPolicy.Properties.Roles.push(role);
+  }
+
+  warnWildcardCustomPolicies(customPolicies, resourceName);
+
+  for (const customPolicy of customPolicies) {
+    validateCustomPolicy(customPolicy, category, resourceName);
+    const policyWithEnv = await replaceEnvForCustomPolicies(customPolicy, envName);
+    if (!policyWithEnv.Effect) policyWithEnv.Effect = iam.Effect.ALLOW;
+    customExecutionPolicy.Properties.PolicyDocument.Statement.push(policyWithEnv);
+  }
+
+  if (service === 'Lambda') {
+    cfnTemplate.Resources.CustomLambdaExecutionPolicy = customExecutionPolicy;
+  }
+  if (service === 'ElasticContainer') {
+    cfnTemplate.Resources.CustomExecutionPolicyForContainer = customExecutionPolicy;
+  }
+
+  await writeCFNTemplate(cfnTemplate, filePath, { templateFormat });
+}
+
+
+//validate the format of actions and ARNs for custom IAM policies
+function validateCustomPolicy(customPolicy: CustomIAMPolicy, category: string, resourceName: string) {
+  const resources = customPolicy.Resource;
+  const actions = customPolicy.Action;
+  const resourceRegex = new RegExp('arn:(aws[a-zA-Z0-9-]*):([a-zA-Z0-9\\-])+:([a-z]{2}(-gov)?-[a-z]+-\\d{1})?:(\\d{12})?:(.*)');
+  const actionRegex = new RegExp('[a-zA-Z0-9]+:[a-z|A-Z|0-9|*]+');
+  const wrongResourcesRegex = [];
+  const wrongActionsRegex = [];
+  let errorMessage = "";
+
+  for (const resource of resources) {
+    if (!(resourceRegex.test(resource) || resource === '*')) {
+      wrongResourcesRegex.push(resource);
+    }
+  }
+
+  for (const action of actions) {
+    if (!actionRegex.test(action)) {
+      wrongActionsRegex.push(action);
+    }
+  }
+
+  const customPoliciesPath = pathManager.getCustomPoliciesPath(category, resourceName);
+
+  if (wrongResourcesRegex.length > 0) {
+    errorMessage += `Invalid custom IAM policy for {function name}. Incorrect "Resource": ${wrongResourcesRegex.toString()}\n Edit ${customPoliciesPath} to fix`;
+  }
+  if (wrongActionsRegex.length > 0) {
+    errorMessage += `Invalid custom IAM policy for {function name}. Incorrect "Action": ${wrongActionsRegex.toString()}\n Edit ${customPoliciesPath} to fix`;
+  }
+
+  if (errorMessage.length > 0) {
+    throw new CustomPoliciesFormatError(errorMessage);
+  }
+}
+
+function validateExistCustomPolicies(customPolicies: CustomIAMPolicies) : Boolean {
+  const ajv = new Ajv();
+  const validatePolicies = ajv.compile(CustomIAMPoliciesSchema);
+
+
+  if (!validatePolicies(customPolicies)) {
+    return false;
+  }
+
+  if (customPolicies.length === 1
+    && customPolicies[0].Action?.length === 0
+    && customPolicies[0].Resource?.length === 0) {
+      return false;
+    }
+
+  return true;
```

Comment:
after addressing the comment about fully typing the json schema, this can all simplify to
```suggestion
  return validatePolicies(customPolicies);
```

Reply:
I am doing this if(customPolicies.length === 1... is because customers could specify nothing in the custom policies. This is to validate if they enter anything, if not, we will just skip adding custom policies

Reply:
But if we validate the schema here at the same time, we could not throw the proper error to customers

Reply:
To eliminate the confusion here, I take out the validatePolicies

Reply:
I think the solution here is to have the JSON schema assert that Action and Resource must have length > 0 but allow a top level array length of 0. Then this logic can be:
```js
return validatePolicies(customPolicies) && customPolicies.length > 0
```

Reply:
In default, the customPolicies.length === 1, there are an empty array of Action and an empty array of Resource.